apps/microtvm/zephyr/demo_runtime/src/main.c

```diff
@@ -22,108 +22,145 @@
+void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *esf) {
+#ifdef CONFIG_LED
+  gpio_pin_set(led0_pin, LED0_PIN, 1);
+#endif
+  for (;;) ;
+}
+
+// Used by TVM when a message needs to be formatted.
+size_t TVMPlatformFormatMessage(char* out_buf, size_t out_buf_size_bytes,
+                                const char* fmt, va_list args) {
   return vsnprintk(out_buf, out_buf_size_bytes, fmt, args);
 }
 
+// Used by TVM when an abort operation occurs.
 void TVMPlatformAbort(tvm_crt_error_t error) {
   sys_reboot(SYS_REBOOT_COLD);
-  for (;;)
-    ;
+#ifdef CONFIG_LED
+  gpio_pin_set(led0_pin, LED0_PIN, 1);
+#endif
+  for (;;) ;
 }
 
-K_MEM_POOL_DEFINE(tvm_memory_pool, 64, 1024, 120, 4);
+// Used by TVM to generate random data.
+tvm_crt_error_t TVMPlatformGenerateRandom(uint8_t* buffer, size_t num_bytes) {
+  uint32_t random;  // one unit of random data.
 
-tvm_crt_error_t TVMPlatformMemoryAllocate(size_t num_bytes, DLContext ctx, void** out_ptr) {
-  *out_ptr = k_mem_pool_malloc(&tvm_memory_pool, num_bytes);
-  return (*out_ptr == NULL) ? kTvmErrorPlatformNoMemory : kTvmErrorNoError;
-}
+  // Fill parts of `buffer` which are as large as `random`.
+  size_t num_full_blocks = num_bytes / sizeof(random);
+  for (int i = 0; i < num_full_blocks; ++i) {
+    random = sys_rand32_get();
+    memcpy(&buffer[i * sizeof(random)], &random, sizeof(random));
+  }
+
+  // Fill any leftover tail which is smaller than `random`.
+  size_t num_tail_bytes = num_bytes % sizeof(random);
+  if (num_tail_bytes > 0) {
+    random = sys_rand32_get();
+    memcpy(&buffer[num_bytes - num_tail_bytes], &random, num_tail_bytes);
+  }
 
-tvm_crt_error_t TVMPlatformMemoryFree(void* ptr, DLContext ctx) {
-  k_free(ptr);
   return kTvmErrorNoError;
 }
 
-uint32_t g_utvm_start_time;
-
 #define MILLIS_TIL_EXPIRY 200
 #define TIME_TIL_EXPIRY (K_MSEC(MILLIS_TIL_EXPIRY))
 K_TIMER_DEFINE(g_utvm_timer, /* expiry func */ NULL, /* stop func */ NULL);
 
+uint32_t g_utvm_start_time;
 int g_utvm_timer_running = 0;
 
-#ifdef CONFIG_LED
-/* The devicetree node identifier for the "led0" alias. */
-#define LED0_NODE DT_ALIAS(led0)
-
-#define LED0 DT_GPIO_LABEL(LED0_NODE, gpios)
-#define PIN DT_GPIO_PIN(LED0_NODE, gpios)
-#define FLAGS DT_GPIO_FLAGS(LED0_NODE, gpios)
-
-static struct device* led_pin;
-#endif  // CONFIG_LED
-
+// Used to start system timer.
 tvm_crt_error_t TVMPlatformTimerStart() {
   if (g_utvm_timer_running) {
     TVMLogf("timer already running");
-    return kTvmErrorPlatformTimerBadState;
+    return kTvmErrorSystemErrorMask | 1;
   }
 
 #ifdef CONFIG_LED
-  gpio_pin_set(led_pin, PIN, 1);
+  gpio_pin_set(led0_pin, LED0_PIN, 1);
 #endif
   k_timer_start(&g_utvm_timer, TIME_TIL_EXPIRY, TIME_TIL_EXPIRY);
   g_utvm_start_time = k_cycle_get_32();
   g_utvm_timer_running = 1;
   return kTvmErrorNoError;
 }
 
-tvm_crt_error_t TVMPlatformTimerStop(double* elapsed_time_seconds) {
+// Used to stop system timer.
+tvm_crt_error_t TVMPlatformTimerStop(double* res_us) {
```

Comment:
I think this is supposed to be in seconds now. it used to be in us, but I changed it in #7227